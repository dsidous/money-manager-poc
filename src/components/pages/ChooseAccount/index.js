import React, { PureComponent } from 'react';
import { View, TouchableOpacity, Text, Image, FlatList } from 'react-native';
import { connect } from 'react-redux';
import { StyledText, StyledButton } from '../../atoms';
import { FETCH_BANK_ACCOUNT_DATA, SELECTED_BANK_ACCOUNT_DATA } from "../../../store/constants";
import bankImages from '../../../assets/images/banklogos/index';
import PropTypes from 'prop-types';

const styles = {
  container: {
    flex: 1,
    marginTop: 20,
    paddingHorizontal: 25,
  },
  viewDefaultStyle: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: "center",
    alignItems: "center",
    padding: 20,
    marginVertical: 10,
    backgroundColor: '#fff',
    borderColor: '#D3D3D3',
    borderWidth: 1,
    borderRadius: 5
  },
  accountSelectedStyle: { backgroundColor: '#ccebff', borderColor: '#0000FF' },
  imageStyle: { width: 50, height: 50, flex: 1, marginRight: 30, resizeMode: 'contain' },
  childViewStyle: { flex: 3, flexDirection: 'column' },
  textStyle: { marginBottom: 10 },
  bottomView: { padding: 10, margin: 20, alignItems: 'center' },
}
class ChooseAccount extends PureComponent {

  componentDidMount() {
    this.props.dispatch({ type: FETCH_BANK_ACCOUNT_DATA })
  }

  renderItem = ({ item }) => {
    let selected = '';

    if (this.props.selectedBankAccount && this.props.selectedBankAccount.accountNumber === item.accountNumber) {
      selected = styles.accountSelectedStyle;
    }
    return (
      <TouchableOpacity
        onPress={() => this.props.dispatch({ type: SELECTED_BANK_ACCOUNT_DATA, payload: item })}
        data-test="account-list-item"
      >
        <View style={[styles.viewDefaultStyle, selected]}>
          <Image source={bankImages[this.props.selectedBank.bankName]} style={styles.imageStyle}></Image>
          <View style={styles.childViewStyle}>
            <Text style={styles.textStyle}>
              {this.props.selectedBank.bankName}
            </Text>
            <Text style={styles.textStyle}>
              {item.accountType}
            </Text>
            <Text style={styles.textStyle}>
              {item.sortCode}  {item.accountNumber}
            </Text>
          </View>
        </View>
      </TouchableOpacity>
    )
  }

  render() {
    const { navigation } = this.props;
    return (
      <View style={styles.container}>
        <StyledText size='large' style={{ marginVertical: 15 }}>
          Select your bank account
        </StyledText>
        <StyledText style={{ marginBottom: 15 }}>
          We found three accounts in your name.
          Which account do you use on a daily
          basis?
        </StyledText>
        <FlatList
          data={this.props.bankAccountList}
          extraData={this.props}
          renderItem={this.renderItem}
          keyExtractor={(item) => item.accountNumber}
        />
        <View style={styles.bottomView}>
          <StyledButton
            disabled={!this.props.selectedBankAccount} style={styles.continueButton}
            onPress={() => this.props.navigation.navigate('Dashboard')}
            data-test="nextButton"
          >
            Select & Continue
          </StyledButton>
        </View>
      </View>
    );
  }
}

const mapStateToProps = (state) => ({
  selectedBank: state.bank.selected,
  bankAccountList: state.bank.accountList,
  selectedBankAccount: state.bank.selectedAccount
})

export default connect(mapStateToProps)(ChooseAccount);

ChooseAccount.propTypes = {
  selectedBankAccount: PropTypes.object,
  bankAccountList: PropTypes.array,
}

