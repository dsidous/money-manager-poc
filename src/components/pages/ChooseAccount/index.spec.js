import React from 'react';
import { shallow, configure } from 'enzyme';
import configureStore from 'redux-mock-store';
import Adapter from 'enzyme-adapter-react-16';
import TestRenderer from 'react-test-renderer';
import ChooseAccount from './';

configure({ adapter: new Adapter() });

describe('choose account', () => {
  const initialState = {
    bank: {
      selected: {
        bankName: 'bank1',
        logoImage: 'bank1logo',
      },
      accountList: [
        {
          accountType: 'account1',
          sortCode: 'sortcode1',
          accountNumber: 'accNumber1',
        },
        {
          accountType: 'account2',
          sortCode: 'sortcode2',
          accountNumber: 'accNumber2',
        },
        {
          accountType: 'account3',
          sortCode: 'sortcode3',
          accountNumber: 'accNumber3',
        },
      ],
      selectedAccount: {
        accountType: 'account1',
        sortCode: 'sortcode1',
        accountNumber: 'accNumber1',
      },
    },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const wrapper = shallow(<ChooseAccount store={store} />);

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Selecting first item and navigate to next screen', () => {
    const navigation = { navigate: jest.fn() };
    const testrenderer = TestRenderer.create(<ChooseAccount navigation={navigation} store={store} />);
    const testinstance = testrenderer.root;

    testinstance.findAllByProps({ 'data-test': 'account-list-item' })[0].props.onPress();
    expect(testinstance.findByProps({ 'data-test': 'nextButton' }).props.disabled).toBe(false);

    testinstance.findByProps({ 'data-test': 'nextButton' }).props.onPress();
    expect(navigation.navigate).toHaveBeenCalledWith('Dashboard');
  });
});
