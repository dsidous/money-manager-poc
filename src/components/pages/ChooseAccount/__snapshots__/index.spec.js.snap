// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choose account renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ChooseAccount)
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bankAccountList": Array [
        Object {
          "accountNumber": "accNumber1",
          "accountType": "account1",
          "sortCode": "sortcode1",
        },
        Object {
          "accountNumber": "accNumber2",
          "accountType": "account2",
          "sortCode": "sortcode2",
        },
        Object {
          "accountNumber": "accNumber3",
          "accountType": "account3",
          "sortCode": "sortcode3",
        },
      ],
      "dispatch": [Function],
      "selectedBank": Object {
        "bankName": "bank1",
        "logoImage": "bank1logo",
      },
      "selectedBankAccount": Object {
        "accountNumber": "accNumber1",
        "accountType": "account1",
        "sortCode": "sortcode1",
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bankAccountList": Array [
          Object {
            "accountNumber": "accNumber1",
            "accountType": "account1",
            "sortCode": "sortcode1",
          },
          Object {
            "accountNumber": "accNumber2",
            "accountType": "account2",
            "sortCode": "sortcode2",
          },
          Object {
            "accountNumber": "accNumber3",
            "accountType": "account3",
            "sortCode": "sortcode3",
          },
        ],
        "dispatch": [Function],
        "selectedBank": Object {
          "bankName": "bank1",
          "logoImage": "bank1logo",
        },
        "selectedBankAccount": Object {
          "accountNumber": "accNumber1",
          "accountType": "account1",
          "sortCode": "sortcode1",
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
